services/versioncache/versioncache.go

```diff
@@ -0,0 +1,320 @@
+					// add a small duration ahead, so the next check will happen when agent established a connection.
+					NextCheckAt: time.Now().Add(s.cfg.ServiceFirstCheckDelay),
+				}); err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+type prepareResults struct {
+	ServiceID   string
+	CheckAfter  time.Duration
+	NeedsUpdate bool
+	AgentID     string
+}
+
+// prepareUpdateVersions checks if there is any service that needs software versions update in the cache and
+// shifts the next check time for this service.
+func (s *Service) prepareUpdateVersions() (*prepareResults, error) {
+	results := &prepareResults{CheckAfter: s.cfg.MinCheckInterval}
+
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		filter := models.FindServicesSoftwareVersionsFilter{Limit: pointer.ToInt(1)}
+		servicesVersions, err := models.FindServicesSoftwareVersions(tx.Querier, filter)
+		if err != nil {
+			return err
+		}
+		if len(servicesVersions) == 0 {
+			results.CheckAfter = s.cfg.ServiceCheckInterval
+
+			return nil
+		}
+		if servicesVersions[0].NextCheckAt.After(time.Now()) {
+			results.CheckAfter = time.Until(servicesVersions[0].NextCheckAt)
+			if results.CheckAfter < s.cfg.MinCheckInterval {
+				results.CheckAfter = s.cfg.MinCheckInterval
+			}
+
+			return nil
+		}
+
+		results.NeedsUpdate = true
+		results.ServiceID = servicesVersions[0].ServiceID
+
+		service, err := models.FindServiceByID(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if service.ServiceType != models.MySQLServiceType {
+			return nil
+		}
+
+		pmmAgents, err := models.FindPMMAgentsForService(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if len(pmmAgents) == 0 {
+			return errors.Errorf("pmmAgent not found for service")
+		}
+		results.AgentID = pmmAgents[0].AgentID
+
+		// shift the next check time for this service, so, in case of versions fetch error,
+		// it will not loop in trying, but will continue with other services.
+		nextCheckAt := time.Now().UTC().Add(s.cfg.ServiceCheckInterval)
+		if _, err := models.UpdateServiceSoftwareVersions(tx.Querier, servicesVersions[0].ServiceID,
+			models.UpdateServiceSoftwareVersionsParams{NextCheckAt: &nextCheckAt},
+		); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+func softwareName(s agents.Software) (models.SoftwareName, error) {
+	var softwareName models.SoftwareName
+	switch software := s.(type) {
+	case *agents.Mysqld:
+		softwareName = models.MysqldSoftwareName
+	case *agents.Xtrabackup:
+		softwareName = models.XtrabackupSoftwareName
+	case *agents.Xbcloud:
+		softwareName = models.XbcloudSoftwareName
+	case *agents.Qpress:
+		softwareName = models.QpressSoftwareName
+	default:
+		return "", errors.Errorf("invalid software type %T", software)
+	}
+
+	return softwareName, nil
+}
+
+func (s *Service) updateVersions() (time.Duration, error) {
```

Comment:
so updateVersions updates version only for one service.
It definitely good to write here.

Reply:
done